feature_layers/add-feature-clustering/src/main/java/com/esri/samples/add_feature_clustering/AddFeatureClusteringSample.java

```diff
@@ -0,0 +1,316 @@
+import com.esri.arcgisruntime.portal.PortalItem;
+import com.esri.arcgisruntime.symbology.ClassBreaksRenderer;
+import com.esri.arcgisruntime.symbology.SimpleMarkerSymbol;
+import com.esri.arcgisruntime.symbology.TextSymbol;
+import javafx.application.Application;
+import javafx.geometry.Insets;
+import javafx.geometry.Pos;
+import javafx.scene.Scene;
+import javafx.scene.control.CheckBox;
+import javafx.scene.control.Label;
+import javafx.scene.control.Slider;
+import javafx.scene.layout.Background;
+import javafx.scene.layout.BackgroundFill;
+import javafx.scene.layout.Border;
+import javafx.scene.layout.BorderStroke;
+import javafx.scene.layout.BorderStrokeStyle;
+import javafx.scene.layout.BorderWidths;
+import javafx.scene.layout.CornerRadii;
+import javafx.scene.layout.HBox;
+import javafx.scene.layout.StackPane;
+import javafx.scene.layout.VBox;
+import javafx.scene.paint.Color;
+import javafx.scene.paint.Paint;
+import javafx.scene.text.Font;
+import javafx.stage.Stage;
+
+import com.esri.arcgisruntime.ArcGISRuntimeEnvironment;
+import com.esri.arcgisruntime.mapping.ArcGISMap;
+import com.esri.arcgisruntime.mapping.view.MapView;
+
+public class AddFeatureClusteringSample extends Application {
+
+  private ArcGISMap mMap;
+  private CheckBox mDisplayLabelsCheckbox;
+  private Label mClusterRadiusLabel;
+  private Slider mClusterRadiusSlider;
+  private Label mMaxScaleLabel;
+  private Slider mMaxScaleSlider;
+  private Label mCurrentMapScaleLabel;
+  private FeatureLayer mLayer;
+
+  ClusteringFeatureReduction mClusteringFeatureReduction;
+
+  private MapView mapView;
+  private PortalItem portalItem; // keep loadable in scope to avoid garbage collection
+
+  private final int DEFAULT_CLUSTER_R = 60;
+
+  @Override
+  public void start(Stage stage) {
+
+    try {
+      // create stack pane and application scene
+      StackPane stackPane = new StackPane();
+      Scene scene = new Scene(stackPane);
+
+      // set title, size, and add scene to stage
+      stage.setTitle("Feature Clustering Sample");
+      stage.setWidth(800);
+      stage.setHeight(700);
+      stage.setScene(scene);
+      stage.show();
+
+      // authentication with an API key or named user is required to access basemaps and other location services
+      String yourAPIKey = System.getProperty("apiKey");
+      ArcGISRuntimeEnvironment.setApiKey(yourAPIKey);
+
+      // create a portal and portal item, using the portal and item ID
+      var portal = new Portal("https://www.arcgis.com/");
+      portalItem = new PortalItem(portal, "aa44e79a4836413c89908e1afdace2ea");
+
+      mMap = new ArcGISMap(portalItem);
+
+      // Get the Zurich buildings feature layer once the map has finished loading
+      mMap.addDoneLoadingListener(() -> {
+        if (mMap.getLoadStatus() == LoadStatus.LOADED) {
+          LayerList l = mMap.getOperationalLayers();
+          if (l.isEmpty())
+            return;
+          mLayer = (FeatureLayer) mMap.getOperationalLayers().get(0);
```

Comment:
Could use `l` instead of calling `mMap.getOperationalLayers()` again, and it might be worth giving `l` a more descriptive name.

Reply:
Yes, very true!